src/interaction/FixedLocalTupleComListInteractionTemplate.hpp

```diff
@@ -0,0 +1,408 @@
+#include "Tensor.hpp"
+#include "Particle.hpp"
+#include "FixedLocalTupleList.hpp"
+#include "esutil/Array2D.hpp"
+#include "bc/BC.hpp"
+#include "SystemAccess.hpp"
+#include "storage/Storage.hpp"
+#include "Interaction.hpp"
+#include "types.hpp"
+#include "esutil/Error.hpp"
+
+namespace espressopp {
+    namespace interaction {
+	template < typename _Potential >
+	class FixedLocalTupleComListInteractionTemplate: public Interaction, SystemAccess {
+	    
+	protected:
+	    typedef _Potential Potential;
+	    
+	private:
+	    long unsigned int num_of_part; // total number of particles
+	    long unsigned int num_of_subchain; // total number of subchains
+	    std::map<long unsigned int, Real3D> com_origin;
+	    
+	    int N_Constrain; // number of particles constraining the interaction
+	    int nParticles;  // number of particles in this node
+	    
+	public:
+	    FixedLocalTupleComListInteractionTemplate
+	    (shared_ptr < System > _system,
+	     shared_ptr < FixedLocalTupleList > _fixedtupleList,
+	     shared_ptr < Potential > _potential)
+	      : SystemAccess(_system), fixedtupleList(_fixedtupleList), potential(_potential) {
+		if (! potential) {
+		    LOG4ESPP_ERROR(theLogger, "NULL potential");
+		}
+		System& system = getSystemRef();
+		esutil::Error err(system.comm);
+
+		long unsigned int localN = system.storage -> getNRealParticles();
+		boost::mpi::all_reduce(*system.comm, localN, num_of_part, std::plus<int>());
+
+		FixedLocalTupleList::TupleList::Iterator it(*fixedtupleList);
+		std::vector<Particle*> pList = it->second;
+
+		N_Constrain = pList.size() + 1;
+
+		// Check the length of individula particle list in FixedLocalTuple
+		for (FixedLocalTupleList::TupleList::Iterator it(*fixedtupleList); it.isValid(); ++it) {
+		    std::vector<Particle*> pList = it->second;
+		    
+		    if (pList.size() + 1 != N_Constrain) {
+			std::stringstream msg;
+			msg << "ERROR: Tuple Length is not constant\n";
+			err.setException(msg.str());
+		    }
+		}
+		
+		num_of_subchain = num_of_part/N_Constrain;
+
+		int* pid_list;
+		longint pidK = 0;
+		pid_list = new int[num_of_subchain];
+
+		// Check the particle id in FixedLocalTuple
+		for (FixedLocalTupleList::TupleList::Iterator it(*fixedtupleList); it.isValid(); ++it) {
+		    Particle *p = it->first;
+		    pid_list[pidK] = p->id();
+		    
+		    for(long unsigned int i = 0; i < pidK; i++) {
+			if (pid_list[i]/N_Constrain == pid_list[pidK]/N_Constrain) {
+			    std::stringstream msg;
+			    msg << "ERROR: Particle ID is redundant\n";
+			    err.setException(msg.str());
+			}
+		    }
+		    pidK++;
+		}
+
+		Real3D* totCOM;
+		totCOM = new Real3D[num_of_subchain];
+		Real3D* COM;
+		COM = new Real3D[num_of_subchain];
+		
+		Real3D Li = system.bc->getBoxL();
+		
+		for(long unsigned int i = 0; i < num_of_subchain; i++) {
+		    COM[i] = 0.;
+		    totCOM[i] = 0.;
+		}
+
+		for (FixedLocalTupleList::TupleList::Iterator it(*fixedtupleList); it.isValid(); ++it) {
```

Comment:
this looks like repeat of `computeCOM`, try to not duplicate the code; although here the all_reduce is an issue; maybe `computeCOM` could return `*Real3D`